Initial contribution of HD PowerView binding

beowulfe opened the pull request:
Adds binding for Hunter Douglas PowerView shades.
Closes #659
Signed-off-by: Andy Lintner <dev@beowulfe.com> (github: beowulfe)

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,78 @@
+        
+        <config-description>
+            <parameter name="ipAddress" type="text" required="true">
+                <label>IP</label>
```
please add `<context>network_address</context>`.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,78 @@
+        <channels>
+            <channel id="position" typeId="shade-position"/>
+            <channel id="vane" typeId="shade-vane"/>
+            <channel id="batteryLow" typeId="shade-batteryLow"/>
```
Please use the [existing system channel type](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels) `system.low-battery` instead.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,78 @@
+        <state readOnly="true"></state>
+    </channel-type>
+    
+    <thing-type id="scene">
```
Scenes do not really feel like a good fit for a "Thing" of its own. They are rather a functionality of the hub itself and hence I would expect them to be a channel on the "hub thing".
As they do not even have a state on their own, a single "scene" channel (of type number or string) would actually suffice I would think (although it leaves the question of how best to provide the stateDescription, in case the binding could provide this).

beowulfe replied:
Hmm... the challenge here would be in discovery. The ID of the scene isn't something the user would be aware of. In the Hunter Douglas app, scenes have names, but these aren't, as far as I can tell, discoverable over the local API ( I think they may come from the cloud service ).

kaikreuzer replied:
But [here](https://github.com/openhab/openhab2-addons/pull/660/files#diff-01410ced23fcf46af4001b1035b99202R58), you also seem to retrieve a name for the scene, otherwise you would have no label for your discovery result either?
So instead of creating discovery results for scenes, you could likewise add channels to the hub - it should make no big difference.

beowulfe replied:
Oh, I see what you're saying - a channel that can take a command containing the scene's label? That works, although I feel like it complicates things for the end user some - especially in a future state where items and sitemaps are configured by point-and-click. Maybe when work progresses in that area, the binding will be able to provide an enumeration of possible values... Ok, I'll make the change.

kaikreuzer replied:
That is what I meant with "how best to provide the stateDescription" - in general, the channel could take a scene id as a command and the stateDescription would provide the possible values (1=SceneA, 2=SceneB, etc.) for the use in UIs. But afair, this is currently not easily possible to do in a binding. So a simpler option for now might be to simply add a switch channel for each scene to the hub - this would be very similar to your existing approach, but would not require any additional Things to be defined.

beowulfe replied:
So, I got this working after fixing a bug in ESH around updating bridges, but I can't get the channels to show in PaperUI. It looks like PaperUI is filtering out any channels that don't have a channel type (at least for things that don't have any advanced channels defined): https://github.com/eclipse/smarthome/blob/master/extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/controllers.configuration.js#L457

I'm a bit rusty with Angular, but if that seems like a bug to you, I can try to come up with a fix.

kaikreuzer replied:
1. Yes, it looks like a bug that channels without a channel type are not displayed.
2. This should not matter in your case, because I would expect that all the scene channels HAVE a channel type (which can be statically declared within an XML).

beowulfe replied:
Except the way PaperUI finds the channel type is by looping through all the channels defined on the thing type: https://github.com/eclipse/smarthome/blob/master/extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/controllers.configuration.js#L381

kaikreuzer replied:
ok, THAT is indeed a problem. I actually thought that has already been changed, but apparently not.
I will enter an issue for the start. If you want to try fixing it yourself, have a try, otherwise I will see to have someone else to work on it.

kaikreuzer replied:
FTR: https://github.com/eclipse/smarthome/issues/1129

beowulfe replied:
If you're able to find someone familiar with that code, that seems safer. The more I dig into it, the more scared I am of side-effects - there's a lot going on there (not without purpose - it's just complex functionality).

kaikreuzer replied:
Ok, then leave it for now. But it means you'll probably have to wait 2-3 weeks before it is fixed.

beowulfe replied:
Understood - thanks Kai.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .,
+ lib/jcifs-1.3.18.jar
```
You need this lib for a single method: `NbtAddress.getByName` - This is doing netBIOS name lookups, which isn't trivial as it seems.
Ideally, I would want to see such a feature in `org.eclipse.smarthome.io.net` - but as this lib is LGPL, we do not have a chance to get it to Eclipse :-/. How hard would it be to re-implement this feature, do you know?

beowulfe replied:
I took a look at how Netbios works (purposefully avoiding looking at the jcifs implementation) and it doesn't seem simple. These older windows network protocols tend to be riddled with weird edge cases and workarounds. Will the LGPL license preclude its inclusion in OH as well? I can always drop it and lose discovery of the hub, at an obvious cost to usability.

kaikreuzer replied:
> Will the LGPL license preclude its inclusion in OH as well?

No, in OH we can live with it. But I think that the NetBIOS name lookup is useful in other bindings as well (@kgoderis, weren't you also checking for that manually somewhere?), so instead of including this functionality in a binding, it would rather want to have it in the framework to be shared. And for this, ESH would be the best place to put it. Alternatively, I would put it in openhab-core, but this is only second best...

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.hdpowerview,
+ org.openhab.binding.hdpowerview.handler
+Require-Bundle: com.google.guava,
```
please do not use Require-Bundle directive!

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Export-Package: org.openhab.binding.hdpowerview,
+ org.openhab.binding.hdpowerview.handler
+Require-Bundle: com.google.guava,
+ org.openhab.core.compat1x,
```
No, an ESH-compatible binding must not have any dependencies on compat1x!

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewHubDiscoveryService.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    protected void startScan() {
+        try {
+            NbtAddress address = NbtAddress.getByName(HDPowerViewBindingConstants.NETBIOS_NAME);
```
Is this potentially a long-running operation that could go into timeouts? If so, this should be scheduled as a job, so that the method can return asap.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewHubDiscoveryService.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        startScan();
```
A background discovery should run continuously and react on incoming packets/announcements.
This binding does not really support background discovery as it has to manually scan for the device.
If the load of such a scan can be neglected, you could actually schedule it (e.g. once a minute) so that the discovery also works if the device is added to the network after the openHAB startup.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewSceneDiscoveryService.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        startScan();
```
same comment as above

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewShadeDiscoveryService.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        startScan();
```
same comment as above

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/AbstractHubbedThingHandler.java:
```diff
@@ -0,0 +1,45 @@
+        }
+        ThingHandler handler = bridge.getHandler();
+        if (!(handler instanceof HDPowerViewHubHandler)) {
+            logger.error("Thing " + getThing().getThingTypeUID().getId() + " belongs to the wrong hub type");
```
Potentially, there isn't a handler registered at the moment when this is called - which might not be an error situation.
So logging as debug might be better here.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/HDPowerViewHubHandler.java:
```diff
@@ -0,0 +1,189 @@
+        webTargets = new HDPowerViewWebTargets(client, config.ipAddress);
+        refreshInterval = config.refresh;
+
+        updateStatus(ThingStatus.OFFLINE);
```
I would remove this and let the poll job figure out whether it is online or offline on its first execution.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/HDPowerViewHubHandler.java:
```diff
@@ -0,0 +1,189 @@
+
+    @Override
+    public void initialize() {
+        HDPowerViewHubConfiguration config = getConfigAs(HDPowerViewHubConfiguration.class);
```
you should check for completeness and correctness of the config before passing it into the next call.

beowulfe replied:
If I add a pattern to the config-description parameter of the bridge-type, is that sufficient? Can I rely on the framework to have performed the validation?

kaikreuzer replied:
You can call `BaseThingHandler.validateConfigurationParameters` to have some basic validation done, but this currently does not check against format pattern.
@tomhoefer: It would make sense to add such a check, would you agree? Btw, when calling `validateConfigurationParameters` within `initialize()`, should we allow the method to throw a `ConfigValidationException` as well (currently you have to catch it in the code yourself)?

tomhoefer replied:
@kaikreuzer I have already implemented the pattern validator but not created a corresponding PR because of missing test cases.... I hope that I have finished these test cases in the beginning of next week

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/HDPowerViewHubHandler.java:
```diff
@@ -0,0 +1,189 @@
+                }
+            }
+        } catch (Exception e) {
+            logger.warn("Could not connect to bridge", e);
```
don't do warn logging, the status update will be logged anyhow.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/HDPowerViewHubHandler.java:
```diff
@@ -0,0 +1,189 @@
+            }
+        } catch (Exception e) {
+            logger.warn("Could not connect to bridge", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, e.getMessage());
```
Are you sure that the exception is because the bridge is offline? If there can be different reasons, better catch different exceptions or remove the status detail and only keep the message.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Import-Package: 
+ javax.ws.rs.client;version="2.0.0",
+ javax.ws.rs.core;version="2.0.1",
+ org.codehaus.jackson.annotate,
```
We have gson - not jackson. Can you change the usage of jax-rs to make use of gson instead?

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/api/responses/Scenes.java:
```diff
@@ -0,0 +1,45 @@
+import org.codehaus.jackson.map.annotate.JsonSerialize;
+import org.codehaus.jackson.map.annotate.JsonSerialize.Inclusion;
+
+import jcifs.util.Base64;
```
I would prefer switching to Java 8 and using `java.util.Base64` instead.

kaikreuzer commented:
Thanks @beowulfe, I am done with my first review!

kaikreuzer commented:
Btw, you should rebase your PR, since you have many changes (deletions) to the autelis binding in your PR.

kaikreuzer commented:
@beowulfe Any updates?

kaikreuzer commented:
@beowulfe It would be great to merge this soon!

kaikreuzer commented:
@beowulfe Another pin, please let us merge this!

beowulfe commented:
@kaikreuzer - sorry, I thought I had added an issue in ESH for the blocker I have on this, but I didn't. I just added it now: eclipse/smarthome#1887

This is preventing me from dynamically adding the channels for scenes.

beowulfe commented:
@kaikreuzer - made the requested changes. Everything working after the fix you made to use BridgeBuilder.

Once you've reviewed, I'll squash the commit.

beowulfe commented:
Oh, github identifies the diff as outdated. Not sure how that happened. I'll still leave it unsquashed so you can see what I changed from last time.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/ESH-INF/thing/thing-types.xml:
```diff
@@ -9,9 +9,10 @@
         
         <config-description>
-            <parameter name="ipAddress" type="text" required="true">
+            <parameter name="ipAddress" type="text" required="true" pattern="^\d{1,3}\.\d{1,3}\.\d{1,3}$">
```
Isn't that a bit too restrictive? It does not allow to enter a hostname instead and what about IPv6 addresses?

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewHubDiscoveryService.java:
```diff
@@ -26,33 +29,53 @@
+            backgroundFuture.cancel(true);
+            backgroundFuture = null;
+        }
+        backgroundFuture = scheduler.scheduleAtFixedRate(scanner, 0, 60, TimeUnit.SECONDS);
```
scheduleWithFixedDelay is usually better, since it avoids jobs piling up (e.g. while being in a breakpoint while debugging).

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/discovery/HDPowerViewSceneDiscoveryService.java:
```diff
@@ -33,39 +34,59 @@
+            backgroundFuture.cancel(true);
+            backgroundFuture = null;
+        }
+        backgroundFuture = scheduler.scheduleAtFixedRate(scanner, 0, 60, TimeUnit.SECONDS);
```
scheduleWithFixedDelay is usually better, since it avoids jobs piling up (e.g. while being in a breakpoint while debugging).

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/handler/HDPowerViewHubHandler.java:
```diff
@@ -58,31 +66,44 @@
+                try {
+                    webTargets.activateScene(Integer.parseInt(channelUID.getId()));
+                } finally {
+                    // 5 seconds later, toggle it back off - scenes are stateless
```
Err, better don't do this but rather treat it as a known limitation or ask people to add `{ autoupdate="false" }` to their items. The proper solution is to help getting https://github.com/eclipse/smarthome/issues/595 implemented!

kaikreuzer commented:
Thanks @beowulfe, I just added a few last comments before we can merge.

beowulfe commented:
@kaikreuzer - updated the PR

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/README.md:
```diff
@@ -0,0 +1,93 @@
+</table>
+
+## Channels
+### PowerView Shade
```
Could you add an empty line here? This otherwise isn't rendered correctly at http://docs.openhab.org/addons/bindings/hdpowerview/readme.html.

kaikreuzer commented on addons/binding/org.openhab.binding.hdpowerview/README.md:
```diff
@@ -0,0 +1,93 @@
+  <td>Switch</td>
+  <td>Turning this to ON will activate the scene. Scenes are stateless in the PowerView hub - they have no on/off state. Include { autoupdate="false" } on your item configuration to avoid needing to toggle off and on.</td>
+ </tr>
+</table>
```
I just noticed that you skipped adding a ["Full Example"](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue#full-example) section - could you add a simple one?

beowulfe commented:
@kaikreuzer - working on the updated docs. I'm not sure how to express the channel for the scene in the Thing DSL. I'm not able to dynamically add the channel to a Thing that comes from a configuration file. 

If it's not possible, I can just document that, recommend again to use discovery, and give an example of a hub without scenes.

kaikreuzer commented:
>  I'm not able to dynamically add the channel to a Thing that comes from a configuration file.

You do not need to add anything dynamically, what you have already statically defined in the things-file, right?
Note that you can define channels in the things-file, which are not part of the thing-type, so this should be the equivalent to the dynamic channels.

beowulfe commented:
@kaikreuzer - one last issue around manual configuration. How can I know in the thing handler if the Thing is managed or not? Calling updateThing throws an IllegalStateException. I could simply catch this, but it seems like it could be thrown for more reasons than the Thing being read-only (the message actually contains "Most likely because it is read-only."). Ideally I'd avoid trying to perform the update at all if the Thing can't be updated. Any suggestions?

kaikreuzer commented:
This is a bit of an open discussion, see [here](https://www.eclipse.org/forums/index.php?t=msg&th=1072114&goto=1734789&#msg_1734789).
So my suggestion for now is to indeed catch and ignore the exception (living with the possibility that the cause might have been something else).
